client/services/cardKeyService.js

```diff
@@ -1,16 +1,127 @@
+  );
+  return Math.abs(calculatedChecksum % 255) === parseInt(checksum, 16);
+};
+const replies = {
+  '00': 'OK', // eslint-disable-line
+  '01': 'ERROR', // eslint-disable-line
+  '83': 'NO CARD', // eslint-disable-line
+  '87': 'UNKNOWN INTERNAL ERROR', // eslint-disable-line
+  '85': 'UNKNOWN COMMAND', // eslint-disable-line
+  '84': 'RESPONSE ERROR', // eslint-disable-line
+  '82': 'READER TIMEOUT', // eslint-disable-line
+  '90': 'CARD DOES NOT SUPPORT THIS COMMAND', // eslint-disable-line
+  '8f': 'UNSUPPORTED CARD IN NFC WRITE MODE' // eslint-disable-line
+};
+const parseData = response => {
+  const hexValues = [];
+  for (let i = 0; i < response.length; i += 1) {
+    hexValues.push((response[i] < 16 ? '0' : '') + response[i].toString(16));
+  }
+  const stationId = hexValues[1];
+  const length = hexValues[2];
+  const status = hexValues[3];
+  const flag = hexValues[4];
+  const data = hexValues.slice(5, hexValues.length - 1);
+  const checksum = hexValues[hexValues.length - 1];
+  const valid = validate([stationId, length, status, flag, ...data], checksum);
+  return {
+    valid: valid,
+    status: replies[status],
+    data: valid && replies[status] === 'OK' ? convertUID(data) : data
+  };
+};
+
 module.exports = [
   '$window',
   '$rootScope',
-  function($window, $rootScope) {
+  'alertService',
+  function($window, $rootScope, alertService) {
     $rootScope.$on('$routeChangeStart', function() {
-      angular.element($window).unbind('message');
+      angular.element($window).unbind('click');
+      clearInterval($rootScope.serialInterval);
+      // $rootScope.serialWriter.close();
+      // $rootScope.serialReader.close();
     });
 
     return {
-      listen: function(cb) {
-        angular.element($window).bind('message', function(e) {
-          cb(e.data);
-        });
+      listen: async function(cb) {
+        let port;
+        if (!$rootScope.serialWriter || !$rootScope.serialReader) {
+          try {
+            port = await navigator.serial.requestPort({});
+
+            await port.open({ baudrate: 9600 });
+            $rootScope.serialWriter = port.writable.getWriter();
+            $rootScope.serialReader = port.readable.getReader();
+          } catch (e) {
+            alert(
+              'Failed to initialize because of Serial API permission errors. Please navigate to another route to trigger a user action.'
+            );
+            throw e;
+          }
+        }
+
+        let lastTime = 0;
+        let lastData = 0;
+        const onComplete = input => {
+          const { valid, status, data } = parseData(input);
+          if (valid && status == 'OK') {
+            if (data !== lastData || Date.now() - lastTime > 2000) {
+              cb(data);
+              lastTime = Date.now();
+              lastData = data;
+            }
+          }
+        };
+
+        const readResult = async () => {
+          let message = [];
```

Comment:
Can theoretically be a `const` right, like `hexValues`? 🙃  Not that it makes any sense tho #JS